Mathlib/GroupTheory/Subgroup/Basic.lean

```diff
@@ -0,0 +1,3736 @@
+@[to_additive "Auxiliary definition used to define `liftOfRightInverse`"]
+def liftOfRightInverseAux (hf : Function.RightInverse f_inv f) (g : G₁ →* G₃) (hg : f.ker ≤ g.ker) :
+    G₂ →* G₃ where
+  toFun b := g (f_inv b)
+  map_one' := hg (hf 1)
+  map_mul' := by
+    intro x y
+    rw [← g.map_mul, ← mul_inv_eq_one, ← g.map_inv, ← g.map_mul, ← g.mem_ker]
+    apply hg
+    rw [f.mem_ker, f.map_mul, f.map_inv, mul_inv_eq_one, f.map_mul]
+    simp only [hf _]
+#align monoid_hom.lift_of_right_inverse_aux MonoidHom.liftOfRightInverseAux
+#align add_monoid_hom.lift_of_right_inverse_aux AddMonoidHom.liftOfRightInverseAux
+
+@[to_additive (attr := simp)]
+theorem liftOfRightInverseAux_comp_apply (hf : Function.RightInverse f_inv f) (g : G₁ →* G₃)
+    (hg : f.ker ≤ g.ker) (x : G₁) : (f.liftOfRightInverseAux f_inv hf g hg) (f x) = g x := by
+  dsimp [liftOfRightInverseAux]
+  rw [← mul_inv_eq_one, ← g.map_inv, ← g.map_mul, ← g.mem_ker]
+  apply hg
+  rw [f.mem_ker, f.map_mul, f.map_inv, mul_inv_eq_one]
+  simp only [hf _]
+#align
+  monoid_hom.lift_of_right_inverse_aux_comp_apply
+  MonoidHom.liftOfRightInverseAux_comp_apply
+#align
+  add_monoid_hom.lift_of_right_inverse_aux_comp_apply
+  AddMonoidHom.liftOfRightInverseAux_comp_apply
+
+/-- `liftOfRightInverse f hf g hg` is the unique group homomorphism `φ`
+
+* such that `φ.comp f = g` (`MonoidHom.liftOfRightInverse_comp`),
+* where `f : G₁ →+* G₂` has a RightInverse `f_inv` (`hf`),
+* and `g : G₂ →+* G₃` satisfies `hg : f.ker ≤ g.ker`.
+
+See `MonoidHom.eq_liftOfRightInverse` for the uniqueness lemma.
+
+```
+   G₁.
+   |  \
+ f |   \ g
+   |    \
+   v     \⌟
+   G₂----> G₃
+      ∃!φ
+```
+ -/
+@[to_additive
+      "`liftOfRightInverse f f_inv hf g hg` is the unique additive group homomorphism `φ`
+      * such that `φ.comp f = g` (`AddMonoidHom.liftOfRightInverse_comp`),
+      * where `f : G₁ →+ G₂` has a RightInverse `f_inv` (`hf`),
+      * and `g : G₂ →+ G₃` satisfies `hg : f.ker ≤ g.ker`.
+      See `AddMonoidHom.eq_liftOfRightInverse` for the uniqueness lemma.
+      ```
+         G₁.
+         |  \\
+       f |   \\ g
+         |    \\
+         v     \\⌟
+         G₂----> G₃
+            ∃!φ
+      ```"]
+def liftOfRightInverse (hf : Function.RightInverse f_inv f) :
+    { g : G₁ →* G₃ // f.ker ≤ g.ker } ≃ (G₂ →* G₃)
+    where
+  toFun g := f.liftOfRightInverseAux f_inv hf g.1 g.2
+  invFun φ := ⟨φ.comp f, fun x hx => (mem_ker _).mpr <| by simp [(mem_ker _).mp hx]⟩
+  left_inv g := by
+    ext
+    simp only [comp_apply, liftOfRightInverseAux_comp_apply, Subtype.coe_mk]
+  right_inv φ := by
+    ext b
+    simp [liftOfRightInverseAux, hf b]
+#align monoid_hom.lift_of_right_inverse MonoidHom.liftOfRightInverse
+#align add_monoid_hom.lift_of_right_inverse AddMonoidHom.liftOfRightInverse
+
+/-- A non-computable version of `MonoidHom.liftOfRightInverse` for when no computable right
+inverse is available, that uses `Function.surjInv`. -/
+@[simp,
```

Comment:
Does this need `to_additive`-izing?

Reply:
this should raise a linter error... Yes, it does.